src/pkg/p2p/preheat/controller/controller.go

```diff
@@ -0,0 +1,277 @@
+	}
+	// Check health before saving
+	f, ok := provider.GetProvider(instance.Vendor)
+	if !ok {
+		return 0, fmt.Errorf("no provider registered with name '%s'", instance.Vendor)
+	}
+	p, err := f(instance)
+	if err != nil {
+		return 0, err
+	}
+
+	status, err := p.GetHealth()
+	if err != nil {
+		instance.Status = provider.DriverStatusUnHealthy
+		log.Errorf("Check health of new instance error: %s; set healthy status to unhealthy", err)
+	} else {
+		instance.Status = status.Status
+	}
+
+	instance.SetupTimestamp = time.Now().Unix()
+
+	return cc.iManager.Save(instance)
+}
+
+// DeleteInstance implements @Controller.DeleteInstance
+func (cc *CoreController) DeleteInstance(id int64) error {
+	return cc.iManager.Delete(id)
+}
+
+// UpdateInstance implements @Controller.UpdateInstance
+func (cc *CoreController) UpdateInstance(id int64, properties models.PropertySet) error {
+	if len(properties) == 0 {
+		return errors.New("no properties provided to update")
+	}
+
+	instance, err := cc.iManager.Get(id)
+	if err != nil {
+		return err
+	}
+
+	if err := properties.Apply(instance); err != nil {
+		return err
+	}
+
+	return cc.iManager.Update(instance)
+}
+
+// PreheatImages implements @Controller.PreheatImages
+func (cc *CoreController) PreheatImages(images ...models.ImageRepository) (CompositePreheatingResults, error) {
+	if len(images) == 0 {
+		return nil, errors.New("no images provided to preheat")
+	}
+
+	// Valid the images
+	for _, img := range images {
+		if !img.Valid() {
+			return nil, fmt.Errorf("%s is not a valid image repository", img)
+		}
+	}
+
+	// Directly dispatch to all the instances
+	// TODO: Use async way in future
+	_, instances, err := cc.iManager.List(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// No instances
+	if len(instances) == 0 {
+		return nil, errors.New("no distribution provider instances")
+	}
+
+	// TODO: submit preheat task
+	results := make(CompositePreheatingResults)
+	return results, nil
+}
+
+// GetInstance implements @Controller.GetInstance
+func (cc *CoreController) GetInstance(id int64) (*providerModels.Instance, error) {
+	return cc.iManager.Get(id)
+}
+
+// Init the distribution providers
+func Init(ctx context.Context) {
+	if DefaultController == nil {
+		if c, err := NewCoreController(ctx); err != nil {
+			log.Fatalf("initialize distribution controller error: %s", err)
+		} else {
+			DefaultController = c
+		}
+	}
+}
+
+// Create a preheating status
+func preheatingStatus(taskID, status string, err error) *provider.PreheatingStatus {
+	return &provider.PreheatingStatus{
+		TaskID: taskID,
+		Status: status,
+		Error:  err.Error(),
+	}
+}
+
+// convert the image to preheat image by adding more required data
+func buildImageData(image models.ImageRepository) (*provider.PreheatImage, error) {
+	extEndpoint, err := config.ExtEndpoint()
+	if err != nil {
+		return nil, err
+	}
+
+	// extURL, err := config.ExtURL()
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	access := []*tk.ResourceActions{
```

Comment:
here we may need more discussions. Similar to the scanning, the pull action may be blocked by the security settings (set at the project configurations, such as signature and vulnerability severity)

Reply:
Let us  discussions later.